### PR TITLE
HSEARCH-4140 Dynamic sharding of the event queue for asynchronous automatic indexing

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net/browse/HSEARCH-4140

I think the easiest way to start reviewing this one is to first have a look at the documentation updates: hopefully, this does a good job of explaining the core principles.

### CODE REVIEW

**Comment by yrodiere** [comment]
Tests against all databases: https://ci.hibernate.org/blue/organizations/jenkins/hibernate-search/detail/PR-2695/10/pipeline

**Comment by yrodiere** [comment]
Looks like we have transaction deadlocks on... you guessed it, SQL Server. I'm investigating.

**Comment by yrodiere** [comment]
I pushed an update; it should be good now.
Tests on all DBs: https://ci.hibernate.org/blue/organizations/jenkins/hibernate-search/detail/PR-2695/10/pipeline

**Review by fax4ever — APPROVED** [review]
Nice job. I've just few comments.
+1 to merge it

File: documentation/src/main/asciidoc/reference/coordination.asciidoc [hunk 1]
```diff
@@ -403,6 +436,40 @@ The default for this property is `100`.
...
++
+The expiration interval must be set to a value 3 times larger than the expiration interval (see above).
++
+Low values (closer to the pulse interval) mean a shorter delay before rebalancing
+when a node abruptly leaves the cluster due to a crash or network failure,
+but increased risk of incorrectly considering an agent as expired.
++
+High values (much larger than the pulse interval) mean a longer delay before rebalancing
+when a node abruptly leaves the cluster due to a crash or network failure,
+but reduced risk of incorrectly considering an agent as expired.
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/reference/coordination.asciidoc:472]: They are not specific to dynamic sharding. Pulses are performed with static sharding too, to detect invalid configuration.

File: mapper/orm-coordination-database-polling/src/main/java/org/hibernate/search/mapper/orm/coordination/databasepolling/event/impl/ClusterTarget.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+import org.hibernate.search.mapper.orm.coordination.databasepolling.cluster.impl.ClusterDescriptor;
+import org.hibernate.search.mapper.orm.coordination.databasepolling.cluster.impl.ShardAssignmentDescriptor;
+import org.hibernate.search.mapper.orm.coordination.databasepolling.logging.impl.Log;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+
+class ClusterTarget {
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	static ClusterTarget create(List<Agent> allAgentsInIdOrder) {
+		Agent[] candidatesInIdOrder = allAgentsInIdOrder.toArray( new Agent[0] );
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply mapper/orm-coordination-database-polling/src/main/java/org/hibernate/search/mapper/orm/coordination/databasepolling/event/impl/ClusterTarget.java:25]: That's premature optimization, and besides, it doesn't even perform better: https://stackoverflow.com/a/29444594/6692043 :)

**Comment by yrodiere** [comment]
Thanks. I addressed your comments; let's wait for CI and merge.

**Pull request merged**